Add more text describing threshold computation

joshuagl opened the pull request:
Add some additional text to each "Check for an arbitrary software attack" section describing threshold computation, in an attempt to help TUF implementers avoid falling into the trap of a naive implementation of threshold counting resulting in incorrect, security affecting, behaviour. 

Further enhance this guidance by recommending, in "File formats", that the signatures list only contain one signature per keyid.

This kind of detail will be easier to add in a much clearer way once we rewrite the workflow to call out to subsections (https://github.com/theupdateframework/specification/issues/121), however I wanted to add this information as soon as possible because we continue to see implementers falling into the same trap:
* [Incorrect threshold signature computation](https://github.com/theupdateframework/tuf/security/advisories/GHSA-pwqf-9h7j-7mv8) in the reference implementation.
* [Incorrect threshold signature computation for new root metadata](https://github.com/theupdateframework/tuf/commit/83ac7be525b733f79a7e9bc573ec580ec835f179) in the reference implementation.
* [Improper uniqueness verification of signature threshold](https://github.com/awslabs/tough/security/advisories/GHSA-5q2r-92f9-4m49) in Tough.
* [Deduplicate signatures with the same key_id](https://github.com/heartsucker/rust-tuf/commit/1f80ec14d3171246568048c69af65ecb028e909e) in rust-tuf.
* [Fix duplicate signature validation](https://github.com/advancedtelematic/aktualizr/commit/2880edc13dccadb634975810575ab6f144e3b95e) in aktualizer (an Uptane implementation).
* [Only consider the count of valid keys to verify thresholds](https://github.com/theupdateframework/go-tuf/commit/3ebc49bbf99db8d705a4759caa2b8c9d582d6031) in go-tuf.
* [Incorrect threshold signature computation](https://github.com/php-tuf/php-tuf/issues/154) in php-tuf.

mnm678 reviewed (dismissed):
LGTM. Thanks for the clarification, hopefully this will help prevent future threshold mistakes.

This is related to another approach to this problem in #155, though the approach there would require a breaking change.

joshuagl commented:
Rebased on the latest master with a new version and date added. Please take a look @trishankatdatadog @mnm678

trishankatdatadog commented on tuf-spec.md:
```diff
@@ -1285,12 +1289,15 @@ it in the next step.
+  metadata file MUST have been signed by: (1) a <a>THRESHOLD</a> of keys
+  specified in the trusted root metadata file (version N), and (2) a
+  <a>THRESHOLD</a> of keys specified in the new root metadata file being
+  validated (version N+1).  When computing the <a>THRESHOLD</a> each
```
```suggestion
  validated (version N+1).  When computing the <a>THRESHOLD</a>, each
```
```suggestion
  validated (version N+1).  When computing the <a>THRESHOLD</a> each
```

trishankatdatadog commented on tuf-spec.md:
```diff
@@ -1285,12 +1289,15 @@ it in the next step.
+  specified in the trusted root metadata file (version N), and (2) a
+  <a>THRESHOLD</a> of keys specified in the new root metadata file being
+  validated (version N+1).  When computing the <a>THRESHOLD</a> each
+  <a>KEY</a> must ONLY contribute one <a>SIGNATURE</a> towards the
```
Right, but what does this mean in practice? Checking only for duplicate keyids, or making sure that differently encoded keys with the same parameters don't slip by? Needs to be clearer.

joshuagl replied:
What do you have in mind here? With the info we have in TUF metadata we can easily check for duplicate keyids. Would could recommend (require?) implementations also check that the same `keyval.public` isn't used by multiple contributing keys?

joshuagl replied:
I've tried to be much more explicit with some edits in fe15850

joshuagl commented on tuf-spec.md:
```diff
@@ -534,6 +534,10 @@ All signed metadata objects have the format:
+Note: The "signatures" list SHOULD only contain one <a>SIGNATURE</a> per
+<a for="role">KEYID</a>. This helps prevent multiple signatures by the same key
+being counted erroneously towards the minimum <a>THRESHOLD</a> indicating valid
+metadata.
```
Added a pointer to the detailed description in fe15850

joshuagl commented:
Almost two years later !? 🙊 I've managed to add some more text in an attempt to address @trishankatdatadog's concerns. I rebased on the latest changes and updated version and date.

With the current version and date in the commit this PR SHOULD follow #272

joshuagl commented:
I'd love to get this PR off my backlog, any chance of some reviews @trishankatdatadog, @mnm678 and @lukpueh ?

lukpueh reviewed (approved):
🚀
